lib/http3/common.c

```diff
@@ -1362,12 +1386,28 @@ void h2o_http3_send_h3_datagrams(h2o_http3_conn_t *conn, uint64_t flow_id, h2o_i
     h2o_quic_schedule_timer(&conn->super);
 }
 
-uint64_t h2o_http3_decode_h3_datagram(h2o_iovec_t *payload, const void *_src, size_t len)
+uint64_t h2o_http3_decode_h3_datagram(h2o_http3_conn_t *conn, h2o_iovec_t *payload, const void *_src, size_t len)
 {
     const uint8_t *src = _src, *end = src + len;
     uint64_t flow_id;
 
-    if ((flow_id = ptls_decode_quicint(&src, end)) != UINT64_MAX)
+    if ((flow_id = ptls_decode_quicint(&src, end)) != UINT64_MAX) {
+        if (conn->peer_settings.h3_datagram_rfc) {
+            // After the quarter stream id you get the context-id which for UDP should be 0 and we point the payload to the byte
+            // after that see comments to RFCs in the h2o_http3_send_h3_datagrams function above)
+            if (*src == 0) {
+                src++;
+            } else {
+                // H3 Context-ID not 0
+                return UINT64_MAX;
```

Comment:
Besides the above suggestion, which means we'd want to return the value of the context id so the caller can decide what to do with the datagram, we also want to address the fact that we're using the peer's H3 settings here to determine how to decode or send the datagrams when, as discussed, we want to base it purely on the information in the associated request. It is easy on the send side because at this stage we know the format since we have the stream (and request) prior to calling `h2o_http3_send_h3_datagrams`. On the decode side, we can only know if the datagram will have a context id if we find the associated stream, but the current decode function does not allow for that. To address the send and decode issues we could do the following:

```
void h2o_http3_send_h3_datagrams(h2o_http3_conn_t *conn, int datagram_format, uint64_t flow_id, h2o_iovec_t *datagrams, size_t num_datagrams)
{
    for (size_t i = 0; i < num_datagrams; ++i) {
        h2o_iovec_t *src = datagrams + i;
        uint8_t buf[quicly_encodev_capacity(flow_id) + 1 + src->len], *p = buf; // The '+ 1' is for Context-ID in case of RFC format
        p = quicly_encodev(p, flow_id);
        if (datagram_format == H2O_DATAGRAM_FORMAT_RFC) {
            p[0] = 0; // H3 Context-ID for UDP frames is 0 (https://www.rfc-editor.org/rfc/rfc9298#section-5)
            p++;
        }
        memcpy(p, src->base, src->len);
        p += src->len;
        ptls_iovec_t payload = ptls_iovec_init(buf, p - buf);
        quicly_send_datagram_frames(conn->super.quic, &payload, 1);
    }

    h2o_quic_schedule_timer(&conn->super);
}
```

So the caller of the above function will obtain the datagram_format from the request (from the stream) and pass it as a parameter.

The single decode function could be split into two calls:
```
uint64_t h2o_http3_h3_datagram_get_flow_id(h2o_http3_conn_t *conn, const void *_src, size_t len)
{
    const uint8_t *src = _src, *end = src + len;
    return ptls_decode_quicint(&src, end);
}

uint8_t h2o_http3_datagram_get_payload_and_context_id(h2o_http3_conn_t *conn, int datagram_format, h2o_iovec_t *payload, const void *_src, size_t len)
{
    const uint8_t *src = _src, *end = src + len;
    uint8_t context_id = 0;

    if (datagram_format == H2O_DATAGRAM_FORMAT_RFC) {
        context_id = *src;
        src++;
     }
     *payload = h2o_iovec_init(src, end - src);
     return context_id;
}
```

To decode the datagram you'd first have to get the flow_id, by calling `h2o_http3_h3_datagram_get_flow_id`, which allows you to get the stream and thus the request which allows you to determine the datagram_format (in the request) which you pass as a parameter to `h2o_http3_datagram_get_payload_and_context_id`.

Comments?

Reply:
Yes, there is a bit of a chicken-and-egg problem for receiving datagrams.

Yes, I think that what you are suggesting is a good idea.  Fortunately the flow id is in the same place in both versions.